src/uct/cuda/cuda_ipc/cuda_ipc_iface.c

```diff
@@ -74,18 +75,58 @@ ucs_status_t uct_cuda_ipc_iface_get_device_address(uct_iface_t *tl_iface,
 static ucs_status_t uct_cuda_ipc_iface_get_address(uct_iface_h tl_iface,
                                                    uct_iface_addr_t *iface_addr)
 {
-    *(pid_t*)iface_addr = getpid();
+    uct_cuda_base_sys_dev_map_t *local = (uct_cuda_base_sys_dev_map_t*)iface_addr;
+    int i;
+
+    local->pid = getpid();
+    local->count = uct_cuda_sys_dev_bus_id_map.count;
+
+    for (i = 0; i < local->count; i++) {
+        local->sys_dev[i] = uct_cuda_sys_dev_bus_id_map.sys_dev[i];
+        local->bus_id[i]  = uct_cuda_sys_dev_bus_id_map.bus_id[i];
+    }
+
     return UCS_OK;
 }
 
 static int uct_cuda_ipc_iface_is_reachable(const uct_iface_h tl_iface,
                                            const uct_device_addr_t *dev_addr,
                                            const uct_iface_addr_t *iface_addr)
 {
-    uct_cuda_ipc_iface_t  *iface = ucs_derived_of(tl_iface, uct_cuda_ipc_iface_t);
+    uct_cuda_ipc_iface_t  *iface        = ucs_derived_of(tl_iface, uct_cuda_ipc_iface_t);
+    uct_cuda_base_sys_dev_map_t *remote = (uct_cuda_base_sys_dev_map_t*)iface_addr;
+    uct_cuda_base_sys_dev_map_t *hash;
+    khiter_t khiter;
+    int khret;
+    int i;
+
+    ucs_recursive_spin_lock(&iface->rem_iface_addr_lock);
+
+    khiter = kh_put(cuda_ipc_rem_iface_addr, &iface->rem_iface_addr_hash,
+                    remote->pid, &khret);
+    if ((khret == UCS_KH_PUT_BUCKET_EMPTY) ||
+        (khret == UCS_KH_PUT_BUCKET_CLEAR)) {
+        hash        = &kh_val(&iface->rem_iface_addr_hash, khiter);
+        hash->count = remote->count;
+        hash->pid   = remote->pid;
+
+        for (i = 0; i < remote->count; i++) {
+            hash->sys_dev[i] = remote->sys_dev[i];
+            hash->bus_id[i]  = remote->bus_id[i];
+            ucs_trace("peer pid %ld sys_dev %u bus_id %u",
+                      (long)hash->pid, (unsigned)hash->sys_dev[i],
+                      (unsigned)hash->bus_id[i]);
+        }
+    } else if (khret == UCS_KH_PUT_KEY_PRESENT) {
+        /* do nothing */
+    } else {
+        ucs_error("unable to use cuda_ipc remote_iface_addr hash");
+    }
+    ucs_recursive_spin_unlock(&iface->rem_iface_addr_lock);
 
     return ((uct_cuda_ipc_iface_node_guid(&iface->super) ==
-             *((const uint64_t *)dev_addr)) && ((getpid() != *(pid_t *)iface_addr)));
+             *((const uint64_t *)dev_addr)) &&
```

Comment:
will extra checks be added later?

Reply:
No further checks here.

Reply:
what is the reason to update hash in this function? How it is it going to be used?

Reply:
@brminich updated the PR description as well. Lmk if it makes sense.

UCT perf_estimate API takes two system devices (local, remote) to provide bandwidth/latency estimates. In the case of cuda_ipc, local GPU's sys_dev and remote GPU's sys_dev. But remote GPU's sys_dev may not match the same bus_id on the local process because of the order in which sys_devices are populated. For this reason, we need a way to interpret remote sys_dev and translate to local sys_dev and use in cuda_ipc perf estimate to check if there are nvlinks between the devices or not. Iface address exchange phase seemed like the most convenient place to add this logic.

Reply:
you mean `uct_iface_estimate_perf` API, right? So, we implicitly assume that `iface_is_reachable` for the corresponding iface (created on needed device id) has to be called before this perf estimate routine? If yes, wouldn't it be better to cache device id in ep creation routine rather than during reachability check?